### PR TITLE
WIP, TST: NumPy unit tests no longer use a network connection

### LINKED ISSUE #11631: TST: DataSource test require to be online
Probably just an oddity and me being a bit annoyed/surprised by it, please go ahead and simply close again, but I was confused that when trying to test a bit on the plane I got failures because I was offline. Those come from DataSource, which possibly should just be deprecated anyway (but I have no idea who might be using it).

### PR DESCRIPTION
**Pull request by tylerjereddy**
An example of a solution that fixes #11631, where a few core devs pointed out that when they develop offline `3` unit tests fail--those unit tests all involve `DataSource`.

**Good things** about the approach in this PR: concise, uses an actual fake TCP socket implementation that is local only, and seems to work on my laptop with no network connection (no failing tests for full suite or with  `python runtests.py -t "numpy/lib/tests/test__datasource.py"`)

**Bad things** about this approach: adds a new testing dependency ([httpretty](https://github.com/gabrielfalcao/HTTPretty)) & that could be enough to shoot it down for something like NumPy; the decorator syntax is yet another thing compounded on top of i.e., the syntactic sugar from `pytest` marks.

Alternatives: using standard library `mock` to cook something up with `Mock` or `MagicMock`, [pytest `monkeypatch` fixture](https://docs.pytest.org/en/latest/monkeypatch.html) (I think we want to avoid that level of "magic" from pytest though).

Notes: `[ci skip]` for now because adding a testing dependency is probably too contentious to justify CI cycles for that at the moment.

### CODE REVIEW

**Comment by rgommers** [comment]
Agreed that the extra testing dependency is not nice. Can you make it an optional one? I.e. just catch the ImportError, and skip the tests if `httpretty` is not available.

**Comment by eric-wieser** [comment]
`test_InvalidHTTP` should be changed too - right now, that test just passes automatically if there's no internet connection, which doesn't seem to be in the spirit of what it is testing for.

**Comment by tylerjereddy** [comment]
Also, the monkeypatch stuff does seem to introduce a lot of duplication with those context managers--the idea is to avoid globally monkeypatching network handling behavior because there can be nasty side effects if not careful about this stuff.

File: numpy/lib/tests/test__datasource.py [hunk 1]
```diff
@@ -102,22 +90,35 @@ def teardown(self):
...
+            try:
+                self.ds.open(url)
+            except IOError as e:
+                # Regression test for bug fixed in r4342.
+                assert_(e.errno is None)
 
     def test_InvalidHTTPCacheURLError(self):
-        assert_raises(URLError, self.ds._cache, invalid_httpurl())
+        assert_raises(URLError, self.ds._cache,
+                      "http://fake.abc.web/site/fake.txt")
```
Comment (tylerjereddy) [inline numpy/lib/tests/test__datasource.py]: this test can't currently handle the more garbled URL I added in

**Comment by QuLogic** [comment]
Oh, sorry, I just noticed this PR, but I've made a similar, but possibly simpler(?) attempt in #11715.

**Comment by tylerjereddy** [comment]
@QuLogic Ok, go for it--I added some comments there & I'll close this so you can take charge of addressing comments from Eric et al.

**Pull request closed without merging**